[Notifier][Firebase] Add Firebase v1 API support

vojtechsmejkal opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Issues        | Fix #51147
| License       | MIT

This PR adds support for the new [v1 API](https://firebase.google.com/docs/reference/fcm/rest).

The main changes were:
- Changed authorization process to use short lived JWT
- Changed DSN shape to provide all information required by new API
- New API endpoint was used
- `FirebaseOptions` were extended to hold params supported by new API

This PR does not break backwards compatibility meaning that new Firebase bridge should behave the same way inside the Symfony app as the old one does right now.

Support for the old API was completely removed as the API has been shut down in 2024 already and attempting to use it throws exceptions.

### Deprecations
- `$token` param in `FirebaseTransport::__construct()` was deprecated - the param no longer exists for the new API
- `AndroidNotification`, `IOSNotification` and `WebNotification` classes were deprecated - new Firebase API allows for all platform specific options to be specified for the message and having options separated by platform limits usability
- If you attempt to use the `FirebaseTransport` with the old DSN format a deprecation notice is triggered to inform the user to upgrade (this has been done mainly to not break apps which have the `FirebaseTransport` configured with the old credentials format)

Commit:
[Notifier][Firebase] Require ext-openssl

New Firebase API uses JWT authorization and OpenSSL is needed for signing it.

Commit:
[Notifier][Firebase] Modify FirebaseOptions target to work with the union field from new API

New API changed the 'to' field into a union field with 3 different types. This adds backwards-compatible way to support targeting this union field in FirebaseOptions.

Commit:
[Notifier][Firebase] Rework Firebase transport to be compatible with the new v1 API

Backwards compatibility was maintained with the old DSN format. A deprecation is triggered when someone tries to create the transport with the old DSN format. Transport created using the old DSN will not be able to send messages and will throw exception for every sent message (this should be consistent with its previous behavior).

Commit:
[Notifier][Firebase] Update tests to reflect move to v1 API

The private key used in tests was generated only for the purposes of these tests and is not connected to anything real.

Commit:
[Notifier][Firebase] Rework FirebaseOptions to better reflect new API shape

Shape of the endpoint input has changed with the new API. Separating options for messages based on the target platform no longer makes sense as all options for all platforms can now be set for the same message.

AndroidNotification, IOSNotification and WebNotification were marked as deprecated and will be removed.

FirebaseOptions is no longer abstract and should be used directly.

Commit:
[Notifier][Firebase] Update README and CHANGELOG

Reflects changes made by transferring to new v1 API.

vojtechsmejkal commented on src/Symfony/Component/Notifier/Bridge/Firebase/FirebaseTransport.php:
```diff
@@ -53,48 +70,90 @@ protected function doSend(MessageInterface $message): SentMessage
         return $sentMessage;
     }
+
+    private function getJwt(): string
```
Generating JWT is quite a heavy operation and for example a messages worker dedicated to sending push notifications would perform this operation many times.

Can I add caching for the generated JWT? If so, should I add `symfony/cache` dependency?

nicolas-grekas replied:
Answered by the merged version: the assertion is memoised on the transport and reused until shortly before it expires, so no `symfony/cache` dependency was needed.

Measured on a 2048-bit key: generating it costs 1.75 ms, of which 0.70 ms is `openssl_pkey_get_private()` and 0.51 ms `openssl_sign()`. With the memoisation that drops to 0.0001 ms per call, and it regenerates once expired.

A PSR-6 pool would have been the wrong tool here anyway: it would write a bearer credential to Redis, Memcached or disk, possibly shared with other applications, to save 1.75 ms once per process per hour.

vojtechsmejkal replied:
Understood, PSR-6 pool wouldn't make sense here, thank you for the feedback!

Commit:
[Notifier][Firebase] Add dependency for symfony/deprecation-contracts

Methraen commented:
> Some [tests on Windows](https://github.com/symfony/symfony/actions/runs/14421339615/job/40444302562?pr=60205) are failing. It seems like `ext-openssl` is not installed on that environment. If my assumption is correct, what is the preferred way to resolve this?
> 
> Rest of the failing tests seem to come from Doctrine bridge which should be unrelated to changes made in this PR.


If im not wrong, only Symfony maintainers can modify the CI pipelines configuration,

@nicolas-grekas would it be okay to enable the missing ext so that the PR can be completed?

Commit:
Exclude Firebase Notifier Bridge from running on windows minimal tests

Running tests there makes no sense since ext-openssl is needed.

vojtechsmejkal commented:
I have rebased my branch and I have also modified `.github/workflows/windows.yaml` to exclude tests from this bridge to run on minimal extensions environment.

All tests have passed now.

@nicolas-grekas does this approach make sense?

cesurapp commented:
A problem that hasn't been solved for 2 years, I can't believe it!! I'm still using the package I wrote 2 years ago, what are you waiting for to merge it??

vojtechsmejkal commented:
I'm still waiting for the review. I'll rebase the branch above 8.1 and solve the minor conflicts.

I would also like to see this PR merged in order to not have to keep local overrides of the Firebase logic in my projects.

If I can help with anything else, please let me know.

Commit:
[Notifier][Firebase] Send messages through the Cloud Messaging v1 API

The legacy FCM endpoint the bridge targeted was shut down, so the transport
now authenticates with a service account, signing a short lived JWT, and posts
to /v1/projects/{projectId}/messages:send. FirebaseOptions carries the target
type and the platform specific blocks the v1 message accepts, which makes the
three per-platform notification classes redundant.

nicolas-grekas commented:
Thanks for your patience @vojtechsmejkal, and for doing what was asked back in May 2025 only to hear nothing since. Rebased on 8.2, squashed and pushed.

**One real bug, in the JWT.** `kid` was in the payload, and it is a JOSE *header* parameter, so Google could not select the signing key from it. Generated a JWT against a throwaway RSA key to confirm:

```
before  header : {"alg":"RS256","typ":"JWT"}
        payload: {...,"kid":"KEY_ID_123"}
after   header : {"alg":"RS256","typ":"JWT","kid":"KEY_ID_123"}
        payload: {"iss":...,"sub":...,"aud":"https://fcm.googleapis.com/","iat":...,"exp":...}
```

The signature verified in both cases, so the `openssl_sign()` side was fine; it would have worked while a service account had a single key and broken on rotation.

**The Windows workflow change is no longer needed.** You excluded this bridge from the minimal-extensions job because `ext-openssl` was missing. Both Windows jobs now write `extension=php_openssl.dll` into `php.ini`, so I dropped that hunk and the PR no longer touches `.github/`.

**The three notification tests were never updated.** `AndroidNotificationTest`, `IOSNotificationTest` and `WebNotificationTest` still asserted the legacy flat payload, so all eight failed. They now carry `#[IgnoreDeprecations]`, since the classes they cover are deprecated here, and their expectations match the v1 shape: `android.notification.channel_id`, `apns.payload.aps`, `webpush.notification`, with the target as `topic` rather than `to`.

Smaller things: the deprecation version strings said `7.3` and are now `8.2`; the CHANGELOG entries moved from a `7.3` section to `8.2` and were shortened to one line each; `ext-openssl` and `symfony/deprecation-contracts` are now declared in the bridge's `composer.json` against the current constraints; the README kept its `Sponsor` section, which the branch had dropped; and there was a `tergetType:` typo in the new README examples.

The bridge is green at 26 tests, php-cs-fixer is clean, and the diff is confined to the Firebase bridge.

One thing worth a second opinion before merge: `FirebaseOptions` defaults `targetType` to `Topic`, so a bare string that used to be a device token in the legacy `to` field is now sent as a topic. That is the right default for the new API, but it is a silent behaviour change for anyone constructing options positionally, so it may deserve a line in the deprecation notes.

nicolas-grekas commented:
Thank you @vojtechsmejkal.

vojtechsmejkal commented:
@nicolas-grekas Thank you for the review and for your feedback! I'm happy the PR was merged and I'll be happy to help Symfony development in the future.
